### ISSUE TITLE
Terraform appears to ignore provider version constraint

### ISSUE THREAD

**Issue by dmrzzz** [hashicorp/terraform#17249]
I'm looking forward to #16835, but in the meanwhile I'm having trouble placing a "pessimistic" version constraint in the provider block (which the documentation implies should work).

Given the following configuration:

```hcl
terraform {
  required_version = "~> 0.11"

  ## future (https://github.com/hashicorp/terraform/issues/16835)
  #required_providers {
  #  aws    = "~> 1.7"
  #}
}

provider "aws" {
  region = "us-east-2"

  # until https://github.com/hashicorp/terraform/issues/16835
  version = "~> 1.7"
}

resource "aws_instance" "example" {
  ami = "ami-2757f631"
  instance_type = "t2.micro"
}
```

I expect that I should end up with 1.7.1 of the aws provider, but instead I get 1.8.0

```bash
$ terraform -version
Terraform v0.11.2

$ terraform init

Initializing provider plugins...
- Checking for available provider plugins on https://releases.hashicorp.com...
- Downloading plugin for provider "aws" (1.8.0)...

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

$ terraform -version
Terraform v0.11.2
+ provider.aws v1.8.0

$ terraform providers
.
└── provider.aws ~> 1.7
```

**Comment by bflad** [comment]
Hi @dmrzzz! You're actually very close, but likely off due to a typo in the documentation that was fixed with #17087 (publishing soon). `version = "~> 1.7.0"` would cover `1.7.0`, `1.7.1`, or any `1.7.x` versions.

**Comment by dmrzzz** [comment]
Ah, that makes sense now.  Thanks!
